Update Flash Attention guidance in "Model acceleration libraries" (#5793)
RadeonOpenCompute/ROCm: docs/how-to/rocm-for-ai/inference-optimization/model-acceleration-libraries.rst

   :alt: Attention module of a large language module utilizing tiling
   :align: center

Installation prerequisites
----------------------------

> Before installing Flash Attention 2, make sure the following are available:

Reviewer: ```suggestion
Before installing Flash Attention 2, ensure the following are available:
```

Reviewer: What does available mean in this case?
Can't you just say that these must be installed before installing FA?

*  ROCm-enabled PyTorch
*  Triton

These can be installed by following the official
`PyTorch installation guide <https://pytorch.org/get-started/locally/>`_.  Alternatively, for a simpler setup, you can use a preconfigured
`ROCm PyTorch Docker image <https://github.com/ROCm/rocm-docker-images>`_, which already includes the required libraries.

Installing Flash Attention 2 
----------------------------

> `Flash Attention <https://github.com/Dao-AILab/flash-attention>`_ supports two backend implementation on AMD GPUs.

Reviewer: ```suggestion
`Flash Attention <https://github.com/Dao-AILab/flash-attention>`_ supports two backend implementations on AMD GPUs.
```

*  `Composable Kernel (CK) <https://github.com/ROCm/composable_kernel>`_ - the default backend
*  `OpenAI Triton <https://github.com/triton-lang/triton>`_ - an alternative backend 

You can switch between these backends using the environment variable FLASH_ATTENTION_TRITON_AMD_ENABLE:

FLASH_ATTENTION_TRITON_AMD_ENABLE="FALSE"
→ Use Composable Kernel (CK) backend (FlashAttention 2)

FLASH_ATTENTION_TRITON_AMD_ENABLE="TRUE"
→ Use OpenAI Triton backend (FlashAttention 2)

To install Flash Attention 2, use the following commands:

.. code-block:: shell

   git clone https://github.com/Dao-AILab/flash-attention.git
   cd flash-attention/
   pip install ninja

   # To install the CK backend flash attention
   python setup.py install 

   # To install the Triton backend flash attention
   FLASH_ATTENTION_TRITON_AMD_ENABLE="TRUE" python setup.py install 

   # To install both CK and Triton backend flash attention
   FLASH_ATTENTION_TRITON_AMD_ENABLE=TRUE && FLASH_ATTENTION_SKIP_CK_BUILD=FALSE python setup.py install

> For detailed installation instructions, refer to the `Flash Attention <https://github.com/Dao-AILab/flash-attention>`_

Reviewer: ```suggestion
For detailed installation instructions, see `Flash Attention <https://github.com/Dao-AILab/flash-attention>`_.
```

Benchmarking Flash Attention 2 
------------------------------

> In the `flash-attention/benchmarks/` directory, you can find benchmark scripts to evaluate the performance of Flash Attention 2

Reviewer: ```suggestion
Benchmark scripts to evaluate the performance of Flash Attention 2 are stored in the `flash-attention/benchmarks/` directory.
```

To benchmark the CK backend 

.. code-block:: shell

   cd flash-attention/benchmarks
   pip install transformers einops ninja

   python3 benchmark_flash_attention.py 

To benchmark the Triton backend

.. code-block:: shell

   FLASH_ATTENTION_TRITON_AMD_ENABLE="TRUE" python3 benchmark_flash_attention.py

Using Flash Attention 2 
-----------------------

.. code-block:: python

   import torch
   from transformers import AutoModelForCausalLM, AutoTokenizer

>    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

Reviewer: There are several references to cuda in the code - this is good, right?

   model_name = "NousResearch/Llama-3.2-1B"

   tokenizer = AutoTokenizer.from_pretrained(model_name, dtype=torch.bfloat16, use_fast=False)
   inputs = tokenizer('Today is', return_tensors='pt').to(device)

   model_eager = AutoModelForCausalLM.from_pretrained(model_name, dtype=torch.bfloat16, attn_implementation="eager").cuda(device)
   model_ckFAv2 = AutoModelForCausalLM.from_pretrained(model_name, dtype=torch.bfloat16, attn_implementation="flash_attention_2").cuda(device)
   model_eager.generation_config.pad_token_id = model_eager.generation_config.eos_token_id
   model_ckFAv2.generation_config.pad_token_id = model_ckFAv2.generation_config.eos_token_id

   print("eager\n GQA: ", tokenizer.decode(model_eager.generate(**inputs, max_new_tokens=22)[0], skip_special_tokens=True, do_sample=False, num_beams=1))
   print("ckFAv2\n GQA: ", tokenizer.decode(model_ckFAv2.generate(**inputs, max_new_tokens=22)[0], skip_special_tokens=True, do_sample=False, num_beams=1))

The outputs from eager mode and FlashAttention-2 are identical, although their performance behavior differs.

.. code-block:: shell

   eager

>    GQA:  Today is the 10th anniversary of the 9/11 attacks. I remember that day like it was yesterday.

Reviewer: We sure we want to use something that might be political here (9/11)?
